Folia support

lucko opened the pull request:
Adds support for Paper's regionised/multithreading Folia project.

As you can see, there were fairly minimal changes required to get LuckPerms working, so I don't expect many bugs/issues. It has been lightly tested by me. :)

# Test build

There is a jar you can download from Jenkins: https://ci.lucko.me/job/LuckPerms-Folia/

lucko committed:
Folia support

lucko commented:
## Test build

[LuckPerms-Bukkit-5.4.73.jar.zip](https://github.com/LuckPerms/LuckPerms/files/11130649/LuckPerms-Bukkit-5.4.73.jar.zip)

You'll need to download it, unzip, then add to the plugins directory.

masmc05 commented:
https://mclo.gs/URNzDdt
we don't know what to blame, folia version or that we host the temp server with folia far from main server with the database

This happened randomly, and after that it started spamming a lot with that error, and kicked everyone out, before that it worked 4 days without issues, and after a restart it started to work again without issues

Feels like for some reason it didn't want to reconnect, after some issue happened

pontaoski commented:
Needs rebase, but this seems to work for me on 1.19 Folia

lucko commented:
Rebased and added support for MC 1.20.1 :) [LuckPerms-Bukkit-5.4.96.jar.zip](https://github.com/LuckPerms/LuckPerms/files/11859730/LuckPerms-Bukkit-5.4.96.jar.zip)

pontaoski commented:
`bukkit/src/main/java/me/lucko/luckperms/bukkit/messaging/PluginMessageMessenger.java:63` should be updated to Folia; the Folia code would look something like `runAtFixedRate` on
```java
server.getAsyncScheduler().runAtFixedRate(plugin, task -> {
    Collection<? extends Player> players = PluginMessageMessenger.this.plugin.getBootstrap().getServer().getOnlinePlayers();
    Player p = Iterables.getFirst(players, null);
    if (p == null) {
        return;
    }

    p.sendPluginMessage(PluginMessageMessenger.this.plugin.getLoader(), CHANNEL, buf);
    task.cancel();
}, 20L, 2000L, TimeUnit.MILLISECONDS);
```

powercasgamer commented on bukkit/src/main/java/me/lucko/luckperms/bukkit/LPBukkitBootstrap.java:
```diff
@@ -311,4 +313,13 @@ private static boolean checkIncompatibleVersion() {
+
+    private static boolean isFolia() {
+        try {
+            Class.forName("io.papermc.paper.threadedregions.scheduler.RegionScheduler");
```
```suggestion
            Class.forName("io.papermc.paper.threadedregions.RegionizedServer");
```

paper api now includes the folia scheduler stuff

xyfwdy commented:
merge plz !

ewof commented:
foliabros..

ewof commented:
.

powercasgamer commented:
> .

don't post useless comments on prs

DxsSucuk commented:
Anything new about this?

securebread commented:
is there a download for 1.20.2?

SquareCodeFX commented:
@lucko how are things going here?

lucko commented:
I have rebased this PR and updated to Folia 1.20.2 - builds are available from Jenkins: https://ci.lucko.me/job/LuckPerms-Folia/

BearLeonLight commented:
What's the progress and when will it be merged?
Currently I'm using it normally on 1.20.4.

iLemon commented:
pls merge

RitaSister commented:
really, when been is merged?

lucko commented:
Rebased and updated to 1.20.6

RitaSister commented:
> Rebased and updated to 1.20.6

build is failed

lucko commented:
> > Rebased and updated to 1.20.6
> 
> build is failed

Thanks, fixed it

Dovias commented:
> Rebased and updated to 1.20.6

Thank you for your efforts on Luckperms plugin, I was planning to maybe give updating to 1.20.6 a try, but you were quicker :1st_place_medal:  :heart:

Warriorrrr commented on bukkit/src/main/java/me/lucko/luckperms/bukkit/listeners/BukkitConnectionListener.java:
```diff
@@ -240,7 +240,7 @@ public void onPlayerQuit(PlayerQuitEvent e) {
         // perform unhooking from bukkit objects 1 tick later.
         // this allows plugins listening after us on MONITOR to still have intact permissions data
-        this.plugin.getBootstrap().getServer().getScheduler().runTaskLater(this.plugin.getLoader(), () -> {
+        this.plugin.getBootstrap().getScheduler().sync(player, () -> {
```
This part here currently does not seem to work on folia, since the player's scheduler gets retired and the task isn't ran.

Aruvelut-123 commented:
is it support 1.21.1? I need to use it because my server is 1.21.1 folia and no luckperms like plugin for that.

Aruvelut-123 commented:
> > 它支持 1.21.1 吗？我需要使用它，因为我的服务器是 1.21.1 作品集，没有像插件这样的运气。
> 
> 完全可以用

好的谢谢

Panda260 commented:
Is there a date or time when Folia will be supported in the official version?

OoLunar commented:
Bump. Was able to rebase just fine on the latest commit

Q2297045667 commented:
is it support 1.21.4? I need it

x1aoren commented:
> is it support 1.21.4? I need it

你试试，应该可以，我的服务器还没有升级到1.21.4。1.21.2可以使用

Q2297045667 commented:
> > is it support 1.21.4? I need it
> 
> 你试试，应该可以，我的服务器还没有升级到1.21.4。1.21.2可以使用

看起来运行的还不错，我还没有深入的测试

It looks to be running very well, I haven't conducted in-depth testing yet

Euphillya commented:
On 1.21.3+, context world don't work

Ownercz commented:
Hi, any chance for merge?

kaxonomy commented:
Bump, please merge

BNTFryingPan commented:
this pr does not need to be bumped, nor does it need to be merged in order to use luckperms on folia. if you are running a folia server, see the link for test builds in the first post

Q2297045667 commented:
The latest version looks good.Will he be able to merge it within this year?
@lucko

Q2297045667 commented:
@lucko Can you update the version of LuckPerms?

zhaochenkai1 commented:
the 1.21.8 folia server don't work

Mailootje commented:
Does it work on 1.21.8?

RoyvanEmpel commented:
Is there any reason (except for the outdated version) that this isn't beeing implemented? Is something different planned? Or has it just not been done yet.

rafalohaki commented:
works nice
